server/src/main/java/org/elasticsearch/index/IndexModule.java

```diff
@@ -144,6 +145,7 @@ public final class IndexModule {
     private final IndexSettings indexSettings;
     private final AnalysisRegistry analysisRegistry;
     private final EngineFactory engineFactory;
+    private final SetOnce<CheckedFunction<Directory, Directory, IOException>> indexDirectoryWrapper = new SetOnce<>();
```

Comment:
I wonder if we should have a dedicated `FunctionalInterface` for this rather then more generic function? This might help a lot with finding usages. 

The line below scares me.

Reply:
Sounds like a good idea, I pushed 6b847c97423fa66311b3a09c81ae50e131ebe16b